src/webgpu/api/validation/copyBufferToBuffer.spec.ts

```diff
@@ -0,0 +1,263 @@
+    size: bufferSize,
+    usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+  });
+
+  const errorBuffer = t.expectGPUError('validation', () =>
+    t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.MAP_WRITE | GPUBufferUsage.MAP_READ,
+    })
+  );
+
+  const srcOffset = 0;
+  const dstOffset = 0;
+  const copySize = 8;
+
+  t.TestCopyBufferToBuffer({
+    srcBuffer: errorBuffer,
+    srcOffset,
+    dstBuffer: validBuffer,
+    dstOffset,
+    copySize,
+    isSuccess: false,
+  });
+
+  t.TestCopyBufferToBuffer({
+    srcBuffer: validBuffer,
+    srcOffset,
+    dstBuffer: errorBuffer,
+    dstOffset,
+    copySize,
+    isSuccess: false,
+  });
+});
+
+g.test('buffer_usage')
+  .params([
+    { srcUsage: GPUBufferUsage.COPY_SRC, dstUsage: GPUBufferUsage.COPY_DST, isSuccess: true },
+    { srcUsage: GPUBufferUsage.COPY_DST, dstUsage: GPUBufferUsage.COPY_SRC, isSuccess: false },
+    { srcUsage: GPUBufferUsage.COPY_SRC, dstUsage: GPUBufferUsage.COPY_SRC, isSuccess: false },
+  ] as const)
+  .fn(async t => {
+    const { srcUsage, dstUsage, isSuccess } = t.params;
+
+    const bufferSize = 16;
+    const srcOffset = 0;
+    const dstOffset = 0;
+    const copySize = 8;
+
+    const srcBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: srcUsage,
+    });
+    const dstBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: dstUsage,
+    });
+
+    t.TestCopyBufferToBuffer({
+      srcBuffer,
+      srcOffset,
+      dstBuffer,
+      dstOffset,
+      copySize,
+      isSuccess,
+    });
+  });
+
+g.test('copy_size')
+  .params([
+    { copySize: 0, isSuccess: true },
+    { copySize: 2, isSuccess: false },
+    { copySize: 4, isSuccess: true },
+    { copySize: 5, isSuccess: false },
+    { copySize: 8, isSuccess: true },
+  ] as const)
+  .fn(async t => {
+    const { copySize, isSuccess } = t.params;
+
+    const bufferSize = 16;
+    const srcOffset = 0;
+    const dstOffset = 0;
+
+    const srcBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC,
+    });
+    const dstBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_DST,
+    });
+
+    t.TestCopyBufferToBuffer({
+      srcBuffer,
+      srcOffset,
+      dstBuffer,
+      dstOffset,
+      copySize,
+      isSuccess,
+    });
+  });
+
+g.test('copy_offset')
+  .params([
+    { srcOffset: 0, dstOffset: 0, isSuccess: true },
+    { srcOffset: 2, dstOffset: 0, isSuccess: false },
+    { srcOffset: 4, dstOffset: 0, isSuccess: true },
+    { srcOffset: 5, dstOffset: 0, isSuccess: false },
+    { srcOffset: 8, dstOffset: 0, isSuccess: true },
+    { srcOffset: 0, dstOffset: 2, isSuccess: false },
+    { srcOffset: 0, dstOffset: 4, isSuccess: true },
+    { srcOffset: 0, dstOffset: 5, isSuccess: false },
+    { srcOffset: 0, dstOffset: 8, isSuccess: true },
+    { srcOffset: 4, dstOffset: 8, isSuccess: true },
```

Comment:
nit:
```suggestion
    { srcOffset: 4, dstOffset: 4, isSuccess: true },
```
just to make sure dstOffset allows the smallest allowed alignment